walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,136 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.Iterator;
+
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    private ArrayList<TimeRange> availableTimes = new ArrayList<>();
+
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        availableTimes.clear();
+
+        // If the request is too long, return empty ArrayList
+        if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+            return availableTimes;
+        }
+
+        /* Any time is available during the start. Time ranges will 
+            be dynamically added and removed from the availableTimes 
+            array as the program loops through the events */
+        availableTimes.add(TimeRange.WHOLE_DAY);
+        
+        // If the set of attendees is empty, the entire day is available
+        if (request.getAttendees().isEmpty() && request.getOptionalAttendees().isEmpty()) {
+            return availableTimes;
+        }
+
+        Collection<String> attendees = request.getAttendees();
+        Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+        // Boolean: mandatoryAttendee
+        Map<String, Boolean> attendeesMap = new HashMap<String, Boolean>();
+        for (String attendee : attendees) { 
+            attendeesMap.put(attendee, true); 
+        }
+        for (String attendee : optionalAttendees) { 
+            attendeesMap.put(attendee, false); 
+        }
+
+        System.out.println(attendeesMap.toString());
+
+        /* Events stored in Map to allow to dynamically remove events in loop (so as not to 
+           work on the same event that is shared by multiple attendees) */
+        Map<Event, Boolean> eventsMap = new HashMap<Event, Boolean>();
+        for (Event event : events) { 
+            eventsMap.put(event, true); 
+        }
+
+        Iterator eventsIterator; 
+        for (Map.Entry<String, Boolean> attendeeEntry : attendeesMap.entrySet()) {
+            String attendee = attendeeEntry.getKey();
+
+            eventsIterator = eventsMap.entrySet().iterator();
+            while(eventsIterator.hasNext()) {
+
+                Map.Entry pair = (Map.Entry) eventsIterator.next();
+                if ((Boolean) pair.getValue()) {
```

Comment:
Value here is already a boolean.  Is there a reason to re-cast to boolean?

Reply:
For some reason the compiler sees it as an Object. 
I get this error message without explicitly parsing it to a boolean: 
[78,34] incompatible types: java.lang.Object cannot be converted to boolean
any idea why this is happening?

Reply:
I decided to use a Set instead of Map to store the events. I forgot that the main reason I wanted to use a Map was to be able to remove elements while looping, which I could do with the iterator (so I didn't really need the Boolean value).

Reply:
Java I think since you have a **B**oolean object, you'll need to call [booleanValue](https://docs.oracle.com/javase/8/docs/api/java/lang/Boolean.html#booleanValue--) to pull the value?  Really weird why recasting to **B**oolean would work though.

Another alternative (which I'm not sure if Java would allow) is if you use the primitive **b**oolean instead of the Object on your map